### DISCUSSION TITLE
Taskbar Label Suggestions

### DISCUSSION THREAD

**Discussion by Tellinq** [ramensoftware/windhawk-mods#43] [Mod ideas]
Ok, I actually first should admit, I've never been a bit fan of taskbar labels. At least... not the way Windows approaches it in Windows Vista - 10. Mainly, because of the lack of flexibility provided and enforcing you to "decombine" applications. With Taskbar Labels in Windhawk, this does provide slight flexibility with the ability to change the item widt, while not requiring it to "de-combine" applications.

Here are some suggestions I have to make Taskbar Labels more enticing for me:

1. Adaptive widths (min-max)
While being able to change the maximum width is great, not all apps may benefit due to how short or long the text is.

Here is an example of two applications having a short length, causing wasted space:
![image](https://user-images.githubusercontent.com/59412384/186239789-1bef5861-b873-409f-b6cd-42a23e9e6b53.png)

Some apps may also need a slightly longer width such as the one below:
![image](https://user-images.githubusercontent.com/59412384/186239864-a67e3667-776f-4e32-a925-b015623e4f5a.png)

Of course, some applications just have a simply long label and should be capped
![image](https://user-images.githubusercontent.com/59412384/186799919-60497b09-6484-4b1a-8d14-cde273149894.png)
![image](https://user-images.githubusercontent.com/59412384/186800048-4dbce500-72f4-4e06-acdc-3191b16ea041.png)

There should be some options to 
- Toggle if the item width is a static value or adapts to text width
- Set the minimum and maximum width if an adaptive text width option is enabled (probably around 100-200).
- Not only should these be applied globally, but there should also be a way to individually add applications to override the global values.

2. Only showing taskbar labels when…
- An option to show the taskbar label **only** when the application is focused.
    - If this option is not enabled, another option to not show labels for minimized applications (if even possible to detect).
- Sometimes it may be best to only show taskbar labels when the taskbar is not full. This is an option in Windows 10, although prioritized for combining groups only if the taskbar is not full. I know this is a bit of a limitation right now, but when this limitation is addressed, this should also be an option.

Again, these should also be done to allow at a global and/or individual scale.

3. Ability to change the format of how the taskbar labels shows group amount
Currently, the prefix to group numbers is **[%groupNumber%] %appName%**. 
![image](https://user-images.githubusercontent.com/59412384/186240609-7cecf1c9-a5bc-4e13-9eec-53aed6ee2cec.png)

This should also be a field to customize. For example, I may want **(3) GitHub** or **{4} Files**. Like the last two suggestions, this should also be both a global and individual basis.
![example](https://i.imgur.com/teUn4uO.gif)

Finally, thank you for even implementing this in Windhawk! So far, I enjoy this implementation as it is right now, but I think these suggestions will make this even better.

**Comment by m417z** [comment]
Thank you for the feedback.

I implemented suggestion 3 in version 1.1 of the Taskbar Labels for Windows 11 mod. More details about the new version:
https://ramensoftware.com/windhawk-mod-update-taskbar-labels-for-windows-11

Suggestion 1 is interesting, but I'm afraid that it would be too jumpy for daily usage. I went on to see how difficult it would be to implement, and a basic implementation turned out to be rather simple:
https://gist.github.com/m417z/feabfb22feeb745f876c4a520a5bdff1

![Demonstration](https://user-images.githubusercontent.com/4129781/224508554-86e2c334-c49d-455d-9982-50aefd1eca41.gif)

It's just a quick experiment and not everything works, e.g. it doesn't handle the scenario of too many items on the taskbar. Feel free to play with it and report your experience. Let me know if you need help with the installation.

Suggestion 2 - I'm not sure about showing labels only for the active window, I'm afraid it would be rather jumpy, too. But the good news is that version 1.1 finally implements proper handling for a large amount of taskbar items, so they shrink in size and finally the labels disappear once there's no space left.
